File: composeApp/src/androidMain/kotlin/com/dnfapps/arrmatey/ui/calendar/CalendarMonthGrid.kt
package com.dnfapps.arrmatey.ui.calendar

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextAlign
import com.dnfapps.arrmatey.arr.state.CalendarState
import com.dnfapps.arrmatey.entensions.daysInMonth
import com.dnfapps.arrmatey.shared.MR
import com.dnfapps.arrmatey.utils.mokoString
import kotlinx.datetime.LocalDate

@Composable
fun CalendarMonthGrid(
    currentMonth: LocalDate,
    selectedDate: LocalDate?,
    onDateSelected: (LocalDate) -> Unit,
    state: CalendarState
) {
    val firstDayOfMonth = LocalDate(currentMonth.year, currentMonth.month, 1)
    val daysInMonth = currentMonth.daysInMonth()
    val firstDayOfWeek = firstDayOfMonth.dayOfWeek.ordinal+1

    Column(modifier = Modifier.fillMaxWidth()) {
        Row(modifier = Modifier.fillMaxWidth()) {
            listOf(
                MR.strings.sun, MR.strings.mon, MR.strings.tues, MR.strings.wed,
                MR.strings.thu, MR.strings.fri, MR.strings.sat
            ).forEach { day ->
                Text(
                    text = mokoString(day),
                    modifier = Modifier.weight(1f),
                    textAlign = TextAlign.Center,
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }
        }

        val rows = (daysInMonth + firstDayOfWeek + 6) / 7
        repeat(rows) { row ->
            Row(modifier = Modifier.fillMaxWidth()) {
                repeat(7) { col ->
                    val dayIndex = row * 7 + col - firstDayOfWeek

                    if (dayIndex in 0 until daysInMonth) {
                        val date = LocalDate(currentMonth.year, currentMonth.month, dayIndex + 1)

                        CalendarDayCell(
                            date = date,
                            isSelected = date == selectedDate,
                            items = state.items[date] ?: emptyList(),
                            modifier = Modifier.weight(1f),
                            onClick = { onDateSelected(date) }
                        )
                    } else {
                        Spacer(modifier = Modifier.weight(1f))
                    }
                }
            }
        }
    }
}

File: iosApp/iosApp/Views/Calendar/CalendarMonthGrid.swift
//
//  CalendarMonthGrid.swift
//  iosApp
//
//  Created by Owen LeJeune on 2026-02-09.
//

import SwiftUI
import Shared

struct CalendarMonthGrid: View {
    let currentMonth: Date
    let selectedDate: LocalDate
    let onDateSelected: (LocalDate) -> Void
    let state: CalendarState
    
    private let columns = Array(repeating: GridItem(.flexible(), spacing: 4), count: 7)
    private let weekdaySymbols = [
        MR.strings().sun.localized(), MR.strings().mon.localized(), MR.strings().tues.localized(),
        MR.strings().wed.localized(), MR.strings().thu.localized(), MR.strings().fri.localized(),
        MR.strings().sat.localized()
    ]
    
    var body: some View {
        VStack(spacing: 4) {
            LazyVGrid(columns: columns, spacing: 4) {
                ForEach(weekdaySymbols, id: \.self) { day in
                    Text(day)
                        .font(.caption)
                        .fontWeight(.medium)
                        .foregroundColor(.secondary)
                        .frame(maxWidth: .infinity)
                }
            }
            .padding(.horizontal)
            
            LazyVGrid(columns: columns, spacing: 4) {
                ForEach(calendarDays, id: \.self) { calendarDay in
                    if let date = calendarDay.date {
                        CalendarDayCell(
                            date: date,
                            isSelected: date.isEqual(to: selectedDate),
                            items: state.items[date] ?? [],
                            onClick: {
                                onDateSelected(date)
                            }
                        )
                    } else {
                        Color.clear
                            .aspectRatio(1, contentMode: .fit)
                    }
                }
            }
            .padding(.horizontal)
        }
    }
    
    private var calendarDays: [CalendarDay] {
        guard let monthStart = Calendar.current.date(from: Calendar.current.dateComponents([.year, .month], from: currentMonth)),
              let monthRange = Calendar.current.range(of: .day, in: .month, for: monthStart) else {
            return []
        }
        
        let firstWeekday = Calendar.current.component(.weekday, from: monthStart)
        let leadingEmptyDays = firstWeekday - 1
        
        var days: [CalendarDay] = []
        
        for _ in 0..<leadingEmptyDays {
            days.append(CalendarDay(date: nil))
        }
        
        let components = Calendar.current.dateComponents([.year, .month], from: monthStart)
        for day in monthRange {
            let kotlinDate = LocalDate(
                year: Int32(components.year ?? 2024),
                month: Int32(components.month ?? 1),
                day: Int32(day)
            )
            days.append(CalendarDay(date: kotlinDate))
        }
        
        return days
    }
}

struct CalendarDay: Hashable {
    let date: LocalDate?
}
